prober/kubeconfig.go

```diff
@@ -0,0 +1,92 @@
+package prober
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/ribbybibby/ssl_exporter/config"
+	"gopkg.in/yaml.v2"
+)
+
+type KubeConfigCluster struct {
+	Name    string
+	Cluster KubeConfigClusterCert
+}
+
+type KubeConfigClusterCert struct {
+	CertificateAuthority     string `yaml:"certificate-authority"`
+	CertificateAuthorityData string `yaml:"certificate-authority-data"`
+}
+
+type KubeConfigUser struct {
+	Name string
+	User KubeConfigUserCert
+}
+
+type KubeConfigUserCert struct {
+	ClientCertificate     string `yaml:"client-certificate"`
+	ClientCertificateData string `yaml:"client-certificate-data"`
+}
+
+type KubeConfig struct {
+	Path     string
+	Clusters []KubeConfigCluster
+	Users    []KubeConfigUser
+}
+
+// ProbeKubeconfig collects certificate metrics from kubeconfig files
+func ProbeKubeconfig(ctx context.Context, target string, module config.Module, registry *prometheus.Registry) error {
+	if _, err := os.Stat(target); err != nil {
+		return fmt.Errorf("kubeconfig not found: %s", target)
+	}
+	k, err := ParseKubeConfig(target)
+	if err != nil {
+		return err
+	}
+	k.Path = target
+	for _, c := range k.Clusters {
+		if c.Cluster.CertificateAuthority == "" {
+			continue
+		}
+		// Path is relative to kubeconfig path
+		if !filepath.IsAbs(c.Cluster.CertificateAuthority) {
+			newPath := filepath.Join(filepath.Dir(k.Path), c.Cluster.CertificateAuthority)
+			c.Cluster.CertificateAuthority = newPath
```

Comment:
Relative paths still aren't working for me, I'm getting the same error. I don't think this assignment is actually modifying the value that's passed to `collectKubeconfigMetrics`.

Perhaps you could move this logic into `collectKubeconfigMetrics`?

Reply:
I think the issue is I was trying to modify the slice element from within the loop which won't work. Instead I am creating a new slice using modified value and using that new slice for the main struct.  I am also moving the parsing logic entirely to parsing function so can test the parsing without having to actually have a full cert setup.  I added a test case from your example YAML.